src/models/gamma_poisson_model.py

```diff
@@ -201,30 +206,22 @@ def _kreach(self, k, I, Imax, alpha, beta):
         impressions, of which k are shown.
 
         Args:
-          k:  np.array specifying number of impressions that were seen by the user.
+          k:  (C, ) ndarray specifying number of impressions that were seen
+            by the user.
           I:  Total number of impressions shown to all users.
           Imax: Total size of impression inventory.
           alpha: Parameter of Gamma-Poisson distribution.
           beta: Parameter of Gamma-Poisson distribution.
         Returns:
-          For each k, probability that a randomly chosen user will have an inventory
-          of n impressions, of which k are shown.
+          (C, ) ndarray. For each k, probability that a randomly chosen user
+          will have an inventory of n impressions, of which k are shown.
         """
-        return np.array(
-            [
-                np.sum(
-                    self._knreach(
-                        kv,
-                        np.arange(kv, MAXIMUM_COMPUTATIONAL_FREQUENCY + 1),
-                        I,
-                        Imax,
-                        alpha,
-                        beta,
-                    )
-                )
-                for kv in k
-            ]
-        )
+        k = np.asarray(k)
+        mat = self._knreach(
+            k, np.arange(MAXIMUM_COMPUTATIONAL_FREQUENCY + 1), I, Imax, alpha, beta
+        )  # shape: C x MAXIMUM_COMPUTATIONAL_FREQUENCY
+        upper_triangular = np.triu(mat)
+        return np.sum(upper_triangular, axis=-1)
```

Comment:
Can we not get the same behavior using the `numpy.cumsum`?

The choice of the upper triangular matrix and summing over is generally confusing here IMO, I suggest adding a comment describing what is going on.

Reply:
The changes utilize broadcasting and matrix multiplication with upper triangle matrix to speed up the computation. I am not sure if using `numpy.cumsum` can achieve the same result and performance.

Reply:
@riemanli Here, I only see a sum of with the upper triangular matrix over the last axis. Where is the matrix multiplication with the upper triangular that you mentioned, can you point it out to me?

Reply:
Ah I see. Sorry I misunderstood your points. Please neglect my statement above. Let me answer it again.

> Can we not get the same behavior using the `numpy.cumsum`?

I am assuming you're asking about line 224 only. Given that, I don't think the behavior of `np.sum(axis=-1)` is the same as `np.cumsum(axis=-1)`. Please let me know if I misunderstand something again.

> The choice of the upper triangular matrix and summing over is generally confusing here IMO, I suggest adding a comment describing what is going on.

Sure. Let me explain it here as well. `mat` is a 2-dimension array, so `np.triu` will generate `upper_triangular` as a 2d array from 'mat' with all elements below diagonal in 'mat' being zeros. Then the result is the sum of `upper_triangular` over the last axis and is a 1d array. Ex:
```
mat = 
[
  [1, 2],
  [3, 4],
  [5, 6],
]
upper_triangular = 
[
  [1, 2],
  [0, 4],
  [0, 0],
]
result = 
  [3, 4, 0]
```

Reply:
@riemanli

Thank you for then detailed explanation! Could you please also explain why the new code is equivalent to the previous version? This is not clear to me. I might be missing something here...

Reply:
For sure. I would love to explain the rationale behind it.

If we look at the original code of `_kreach`
```python
return np.array(
    [
        np.sum(
            self._knreach(
                kv,
                np.arange(kv, MAXIMUM_COMPUTATIONAL_FREQUENCY + 1),
                I,
                Imax,
                alpha,
                beta,
            )
        )
        for kv in k
    ]
)
```
and the original code of `_knreach` (note that I switch `k` to `kv` for consistence here):
```python
kprob = scipy.stats.binom.logpmf(kv, n, I / Imax)
return np.exp(kprob + self._logpmf(n, alpha, beta))
```

As a result, the result can be see as the pseudo code below
```python
M = MAXIMUM_COMPUTATIONAL_FREQUENCY    # M > C
k = [1, ..., C]
result = [
    np.sum(
        np.exp(
            scipy.stats.binom.logpmf(kv = 1, np.array([kv = 1, ..., M]), I / Imax) + 
            self._logpmf(np.array([kv = 1, ..., M]), alpha, beta)
        )
    ),
    .
    .
    np.sum(
        np.exp(
            scipy.stats.binom.logpmf(kv = C, np.array([kv = C, ..., M]), I / Imax) + 
            self._logpmf(np.array([kv = C, ..., M]), alpha, beta)
        )
    )
] # size = (C, )
```

So here is what the new change did. Since every element has `np.sum` operation, we can defer it later. Instead of using for-loop, I treat every element as a sum of array with size (1, M). As a result, we can first compute a 2D array `mat`
``` python
mat = [
    [_knreach(kv=1, n=0, other_args), ..., _knreach(kv=1, n=M, other_args)],
    [_knreach(kv=2, n=0, other_args), ..., _knreach(kv=1, n=M, other_args)],
    .
    .
    [_knreach(kv=C, n=0, other_args), ..., _knreach(kv=1, n=M, other_args)]
]
```
Since n is always greater than or equal to kv, we don't need the element with `kv < n`. As a result, we only need the upper triangle part of `mat`. And the last step is taking sum on each row of the upper triangle matrix.

The beauty of this change is that it can utilize the powerful broadcast in `Numpy`, which saves a lot computation time.

Hope that helps!

Reply:
@riemanli 
Awesome, thanks a lot for walking me through the details. This makes sense to me. 

One last questions:
Why is this guaranteed that the input vector k has this structure : `k = [0, ..., C - 1]`?

Reply:
Thanks for putting effort on it! That's a great question. The usage of `_knreach` is to compute the reach frequency distribution from the Gamma Poisson model. The new implementation assumes the values in `k` are consecutive frequency values, i.e. `k = np.arange(min_freq, max_freq + 1)`, which is a fair assumption given that we need the information from every frequency to fit the model well.

If there is any case that we want to have non-consecutive frequency values, e.g. `k = [0, 2, 4, ..., max_freq]`,  we can still generate the distribution from 0 to max_freq and extract the values at [0, 2, 4, ..., max_freq]. The computation won't be wasted too much in this approach as `max_freq` is usually small and broadcasting is fast.

Reply:
@riemanli 

I think your explanation of why you are using the upper triangular matrix is only valid when `k = [0, ..., C - 1]`. Therefore since you mentioned the possibility of `k = np.arange(min_freq, max_freq + 1)` and since it is technically possible for other inputs, I advise against your approach in this PR.

However, if you still prefer to keep it as is, please add to the documentation that `k` has to have a required form and possibly even test for the input to follow that form.

Thanks a lot for following up during this relatively long conversation. I appreciate it.

Reply:
Thank you so much for the thorough review @amiragha! 

@matthewclegg please let me know if you also think we should not have the assumption that `k = np.arange(min_freq, max_freq + 1)` for `_kreach`.

Reply:
Since it's a private function and since Rieman's implementation is consistent with how the function is being used within the class, I think it's ok.

Reply:
Thanks for your reply Matthew.

@amiragha I added some comments and unit tests for the assumption about `k` as per your suggestion. Please let me know if there is any advice. Big thanks.

Reply:
@riemanli Great, thanks a lot for adding the comments as well as walking me through the details.